File: src/main/java/community/whatever/onembackendjava/urlshorten/UrlShortenController.java
package community.whatever.onembackendjava.urlshorten;

import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
public class UrlShortenController {

    private final UrlShortenService urlShortenService;

    @PostMapping("/shorten-url/search")
    public String shortenUrlSearch(@RequestBody String key) {
        return urlShortenService.shortenUrlSearch(key);
    }

    @PostMapping("/shorten-url/create")
    public String shortenUrlCreate(@RequestBody String originUrl) {
        return urlShortenService.shortenUrlCreate(originUrl);
    }
}

File: src/main/kotlin/community/whatever/onembackendkotlin/UrlShortenController.kt
package community.whatever.onembackendkotlin

import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import kotlin.random.Random

@RestController
class UrlShortenController {
    private val shortenUrls = mutableMapOf<String, String>()

    @PostMapping("/shorten-url/search")
    fun shortenUrlSearch(@RequestBody key: String): String {
        return shortenUrls[key] ?: throw IllegalArgumentException("Invalid key")
    }

    @PostMapping("/shorten-url/create")
    fun shortenUrlCreate(@RequestBody originUrl: String): String {
        shortenUrls.forEach { (key, value) ->
            if (value == originUrl) {
                return key
            }
        }
        return Random.nextInt(10000).toString().run {
            shortenUrls[this] = originUrl
            this
        }
    }
}
